pkg/k8shandler/certificates.go

```diff
@@ -2,138 +2,81 @@ package k8shandler
-	secret, err := clusterRequest.GetSecret(secretName)
+func (clusterRequest *ClusterLoggingRequest) extractMasterCerts() (err error) {
+	secret, err := clusterRequest.GetSecret(constants.MasterCASecretName)
 	if err != nil {
 		if errors.IsNotFound(err) {
-			return err
+			return nil
 		}
-		return fmt.Errorf("Unable to extract secret %s to file: %v", secretName, err)
+		return fmt.Errorf("Unable to get secret %s: %v", constants.MasterCASecretName, err)
 	}
-
-	value, ok := secret.Data[key]
-
-	// check to see if the map value exists
-	if !ok {
-		if deprecatedKeys.Has(key) {
-			log.Info("No secret data found. Please be aware but likely not an issue for deprecated keys", "key", key, "deprecatedKeys", deprecatedKeys.List())
-		} else {
-			log.Info("No secret data found", key)
-
+	workDir := utils.GetWorkingDir()
+	for name, value := range secret.Data {
+		if err != utils.WriteToWorkingDirFile(path.Join(workDir, name), value) {
+			return err
 		}
-		return nil
 	}
 
-	return utils.WriteToWorkingDirFile(toFile, value)
+	return nil
 }
 
 func (clusterRequest *ClusterLoggingRequest) writeSecret() (err error) {
 
+	secrets, err := loadFilesFromWorkingDir()
+	if err != nil {
+		return err
+	}
 	secret := NewSecret(
 		constants.MasterCASecretName,
 		clusterRequest.Cluster.Namespace,
-		map[string][]byte{
-			"masterca":  utils.GetWorkingDirFileContents("ca.crt"),
-			"masterkey": utils.GetWorkingDirFileContents("ca.key"),
-		})
-
+		secrets,
+	)
 	utils.AddOwnerRefToObject(secret, utils.AsOwner(clusterRequest.Cluster))
 
 	err = clusterRequest.CreateOrUpdateSecret(secret)
 	if err != nil {
 		return
 	}
-
 	return nil
 }
 
-func (clusterRequest *ClusterLoggingRequest) readSecrets() (err error) {
-
-	for secretName, certMap := range secretCertificates {
-		if err = clusterRequest.extractCertificates(secretName, certMap); err != nil {
-			return
-		}
+func loadFilesFromWorkingDir() (map[string][]byte, error) {
+	workDir := utils.GetWorkingDir()
+	files, err := ioutil.ReadDir(workDir)
+	if err != nil {
+		return nil, err
 	}
-
-	return nil
-}
-
-func (clusterRequest *ClusterLoggingRequest) extractCertificates(secretName string, certs map[string]string) (err error) {
-
-	for secretKey, certPath := range certs {
-		if err = clusterRequest.extractSecretToFile(secretName, secretKey, certPath); err != nil {
-			if errors.IsNotFound(err) {
-				return nil
-			}
-			return
-		}
+	results := map[string][]byte{}
+	for _, f := range files {
+		results[f.Name()] = utils.GetFileContents(path.Join(workDir, f.Name()))
 	}
-
-	return nil
+	return results, nil
 }
 
 //CreateOrUpdateCertificates for a cluster logging instance
 func (clusterRequest *ClusterLoggingRequest) CreateOrUpdateCertificates() (err error) {
-
-	// Pull master signing cert out from secret in logging.Spec.SecretName
-	if err = clusterRequest.readSecrets(); err != nil {
+	mutex.Lock()
```

Comment:
It is not quite enough to lock this mutex here, we also need to lock it around accesses to the cert files, in 
```
[localhost cluster-logging-operator]$ ack GetWorkingDirFileContents pkg
pkg/k8shandler/logstore.go
144:		"elasticsearch.key": utils.GetWorkingDirFileContents("elasticsearch.key"),
145:		"elasticsearch.crt": utils.GetWorkingDirFileContents("elasticsearch.crt"),
146:		"logging-es.key":    utils.GetWorkingDirFileContents("logging-es.key"),
147:		"logging-es.crt":    utils.GetWorkingDirFileContents("logging-es.crt"),
148:		"admin-key":         utils.GetWorkingDirFileContents("system.admin.key"),
149:		"admin-cert":        utils.GetWorkingDirFileContents("system.admin.crt"),
150:		"admin-ca":          utils.GetWorkingDirFileContents("ca.crt"),

pkg/k8shandler/curation.go
172:			"ca":       utils.GetWorkingDirFileContents("ca.crt"),
173:			"key":      utils.GetWorkingDirFileContents("system.logging.curator.key"),
174:			"cert":     utils.GetWorkingDirFileContents("system.logging.curator.crt"),
175:			"ops-ca":   utils.GetWorkingDirFileContents("ca.crt"),
176:			"ops-key":  utils.GetWorkingDirFileContents("system.logging.curator.key"),
177:			"ops-cert": utils.GetWorkingDirFileContents("system.logging.curator.crt"),

pkg/k8shandler/visualization.go
146:			"ca":   utils.GetWorkingDirFileContents("ca.crt"),
147:			"key":  utils.GetWorkingDirFileContents("system.logging.kibana.key"),
148:			"cert": utils.GetWorkingDirFileContents("system.logging.kibana.crt"),
160:	sessionSecret = utils.GetWorkingDirFileContents("kibana-session-secret")
170:			"server-key":     utils.GetWorkingDirFileContents("kibana-internal.key"),
171:			"server-cert":    utils.GetWorkingDirFileContents("kibana-internal.crt"),

pkg/k8shandler/fluentd.go
287:			"ca-bundle.crt": utils.GetWorkingDirFileContents("ca.crt"),
288:			"tls.key":       utils.GetWorkingDirFileContents("system.logging.fluentd.key"),
289:			"tls.crt":       utils.GetWorkingDirFileContents("system.logging.fluentd.crt"),
```

Reply:
This is ok as an interim fix but in the longer term a more maintainable approach is  goroutines+channels (or Communicating Sequential Processes as they were called in the 1960s, sometimes a good idea has to wait for the world to catch up) I can do a short talk about it if that's of interest. The best one-liner from the Go folks is:

    "share memory by communicating, don't communicate by sharing memory".

In summary: don't build mutex walls around thread-safe data structures, only to realize - too late - how threads will end up contending for them. Instead design your process as a set of simple loops with local, contention-free, data; then have them communicate their shared concerns over channels.

Reply:
@syedriko and I discussed the proposed implementation solely as a fix to the current problem with the understanding we will better design cert generation and syncing when moving to replace the script logic with golang lib